File: ptern-typescript/src/codegen/codegen.ts
import {
  collectCaptureValidators,
  compileClassDefinitions,
  compileDefinitions,
  compileExpressionWithRepInfo,
  determineFlags,
  determineIgnoreMatching,
  type RepetitionInfo,
} from "./regex";
import { buildPlan, type SubstitutionPlan } from "./substitution";
import type { ParsedPtern } from "../parser/ast";

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export type CompiledPtern = {
  source: string;
  flags: string;
  ignoreMatching: boolean;
  captureValidators: [string, string][];
  isSubstitutable: boolean;
  ignoreSubstitutionMatching: boolean;
  substitutionPlan: SubstitutionPlan | null;
  repetitionInfo: RepetitionInfo[];
};

export function compile(ptern: ParsedPtern): CompiledPtern {
  const flags = determineFlags(ptern);
  const ignoreMatching = determineIgnoreMatching(ptern.annotations);
  const isSubstitutable = ptern.annotations.some(a => a.name === "substitutable" && a.value);
  const ignoreSubstitutionMatching = ptern.annotations.some(
    a => a.name === "substitutions-ignore-matching" && a.value,
  );
  const classDefs = compileClassDefinitions(ptern.definitions);
  const defs = compileDefinitions(ptern.definitions, classDefs);
  const [source, repetitionInfo] = compileExpressionWithRepInfo(ptern.body, defs, classDefs, 0);
  const captureValidators = collectCaptureValidators(ptern.body, defs, classDefs);
  const defBodies = new Map(ptern.definitions.map(def => [def.name, def.body]));
  const substitutionPlan = isSubstitutable ? buildPlan(ptern.body, defBodies) : null;

  return {
    source,
    flags,
    ignoreMatching,
    captureValidators,
    isSubstitutable,
    ignoreSubstitutionMatching,
    substitutionPlan,
    repetitionInfo,
  };
}

File: ptern-kotlin/src/main/kotlin/io/ptern/codegen/Codegen.kt
package io.ptern.codegen

import io.ptern.parser.ast.ParsedPtern

internal fun compile(ptern: ParsedPtern): CompiledPtern {
    val flags = determineFlags(ptern)
    val ignoreMatching = determineIgnoreMatching(ptern.annotations)
    val isSubstitutable = ptern.annotations.any { it.name == "substitutable" && it.value }
    val ignoreSubstitutionMatching = ptern.annotations.any { it.name == "substitutions-ignore-matching" && it.value }
    val classDefs = compileClassDefinitions(ptern.definitions)
    val defs = compileDefinitions(ptern.definitions, classDefs)
    val emitResult = compileExpressionWithRepInfo(ptern.body, defs, classDefs)
    val captureValidators = collectCaptureValidators(ptern.body, defs, classDefs)
    val defBodies = ptern.definitions.associate { it.name to it.body }
    val substitutionPlan = if (isSubstitutable) buildSubstitutionPlan(ptern.body, defBodies) else null
    val syntheticGroupNames = emitResult.repInfo.map { it.groupName }.toSet()

    return CompiledPtern(
        source = emitResult.source,
        flags = flags,
        ignoreMatching = ignoreMatching,
        captureValidators = captureValidators,
        isSubstitutable = isSubstitutable,
        ignoreSubstitutionMatching = ignoreSubstitutionMatching,
        substitutionPlan = substitutionPlan,
        repetitionInfo = emitResult.repInfo,
        syntheticGroupNames = syntheticGroupNames,
    )
}
